crates/red_knot_python_semantic/src/semantic_index.rs

```diff
@@ -1045,4 +1045,52 @@ match subject:
             vec!["subject", "a", "b", "c", "d", "f", "e", "h", "g", "Foo", "i", "j", "k", "l"]
         );
     }
+
+    #[test]
+    fn for_loops_single_assignment() {
+        let TestCase { db, file } = test_case("for x in a: pass");
+        let scope = global_scope(&db, file);
+        let global_table = symbol_table(&db, scope);
+
+        assert_eq!(&names(&global_table), &["a", "x"]);
+
+        let use_def = use_def_map(&db, scope);
+        let definition = use_def
+            .first_public_definition(global_table.symbol_id_by_name("x").unwrap())
+            .unwrap();
+
+        assert!(matches!(definition.node(&db), DefinitionKind::For(_)));
+    }
+
+    #[test]
+    fn for_loops_simple_unpacking() {
+        let TestCase { db, file } = test_case("for (x, y) in a: pass");
```

Comment:
I think this should create two definitions each for `x` and `y` variable.

Reply:
So I'm currently creating two definitions: one for `x` and one for `y`. Is that what you mean? It's possible I'm not doing that in the "right place", though -- I'm doing it in `add_maybe_multiple_for_definitions()` in `infer.rs`.

I might also be misunderstanding what you mean, as this is my first PR adding new symbols to red-knot

Reply:
Ah, I see. Sorry for not noticing that earlier. 

> So I'm currently creating two definitions: one for `x` and one for `y`. Is that what you mean? It's possible I'm not doing that in the "right place", though -- I'm doing it in `add_maybe_multiple_for_definitions()` in `infer.rs`.

I think what's happening here is that the _types_ are being defined for two definitions but there's still only a single definition. The reason being that the definitions are added at the symbol level and are defined by a unique key (`NodeKey` - a cheaper alternative to the node to reduce the cost of hashing). These are defined in the semantic index builder as is done by `add_definition` call. So, as the definition contains the entire `for` statement, the key would still be the same for each symbol in the target expression. In the example, for both `x` and `y` the definition node key is still the same for statement. 

Each symbol should have a unique definition which, in one way, helps the inference engine to determine the type of that symbol. I've a suggestion in my other comment (https://github.com/astral-sh/ruff/pull/13075/files#r1728920074) on what could that look like.

https://github.com/astral-sh/ruff/blob/342473fd3f1ce40958719910f17154770b641ebd/crates/red_knot_python_semantic/src/semantic_index/builder.rs#L191-L192

Does this help?

Reply:
Thanks, that helps a lot! I see the error now. I'm sort-of surprised that this didn't cause any test failures or panics. (Is there another test I should have added, that would have caught this? Or is it not worth it to worry about that?)

Reply:
> (Is there another test I should have added, that would have caught this? Or is it not worth it to worry about that?)

That's a good question. I don't know if there's any way to test that _specific_ behavior so I wouldn't worry about it right now. This would be something to keep in mind when designing the test infrastructure where we assert that each symbol has a unique definition key or something along that line.

Reply:
I think this is mostly a consequence of doing Definitions without the actual type inference; if you'd had tests for the actual type inference of unpacked names, this would have failed in tests.

Reply:
Oh, another test assertion that I think would have caught this would be to assert that the two definitions you get back for the two symbols are actually different definition IDs? But obviously it's not likely you'd write that assertion without already being aware of the issue.

Reply:
Sorry for triple-posting here :) It just occurred to me that another easy thing we could do that would have caught this up-front (and you could add in this PR) would be to `debug_assert` on the return value of `self.definitions_by_node.insert(...)` inside `add_definition`; I think if we are ever overwriting a previously-inserted `Definition` there, it's a bug.

Reply:
Hmm, I tried adding this debug assertion:

```diff
diff --git a/crates/red_knot_python_semantic/src/semantic_index/builder.rs b/crates/red_knot_python_semantic/src/semantic_index/builder.rs
index 9805d2501..4185765ba 100644
--- a/crates/red_knot_python_semantic/src/semantic_index/builder.rs
+++ b/crates/red_knot_python_semantic/src/semantic_index/builder.rs
@@ -190,8 +190,13 @@ impl<'db> SemanticIndexBuilder<'db> {
             countme::Count::default(),
         );
 
-        self.definitions_by_node
+        let definition_already_existed = self
+            .definitions_by_node
             .insert(definition_node.key(), definition);
+        debug_assert_eq!(
+            definition_already_existed, None,
+            "Should never overwrite an existing definition!"
+        );
         self.current_use_def_map_mut()
             .record_definition(symbol, definition);
```

But it causes a panic in the corpus test (and not due to this PR, I don't think?)

```
checking "/Users/alexw/dev/ruff/crates/red_knot_workspace/resources/test/corpus/58_async_for_dict_comp.py"
thread 'corpus_no_panic' panicked at crates/red_knot_python_semantic/src/semantic_index/builder.rs:196:9:
assertion `left == right` failed: Should never overwrite an existing definition!
  left: Some(Definition { [salsa id]: Id(45d1), file: File { [salsa id]: Id(cc3), path: System("/Users/alexw/dev/ruff/crates/red_knot_workspace/resources/test/corpus/58_async_for_dict_comp.py"), permissions: Some(33188), revision: FileRevision(31781594018582998360096355038), status: Exists, count: Count { ghost: PhantomData<fn(ruff_db::files::File)> } }, file_scope: FileScopeId(2), symbol: ScopedSymbolId(0), node: Comprehension(ComprehensionDefinitionKind { node: AstNodeRef(Comprehension { range: 31..54, target: Tuple(ExprTuple { range: 41..45, elts: [Name(ExprName { range: 41..42, id: Name("k"), ctx: Store }), Name(ExprName { range: 44..45, id: Name("v"), ctx: Store })], ctx: Store, parenthesized: false }), iter: Call(ExprCall { range: 49..54, func: Name(ExprName { range: 49..52, id: Name("gen"), ctx: Load }), arguments: Arguments { range: 52..54, args: [], keywords: [] } }), ifs: [], is_async: true }), first: true }), count: Count { ghost: PhantomData<fn(red_knot_python_semantic::semantic_index::definition::Definition)> } })
 right: None
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    corpus_no_panic
```

So I think I'll leave this for now?

Reply:
Yeah, makes sense; if we're already not keeping this invariant, it's out of scope for this PR. I created an issue to follow up: https://github.com/astral-sh/ruff/issues/13085